# 1. Data Understanding

In [2]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [4]:
df = pd.read_csv('loan_data_2007_2014.csv', index_col = 0)
df.sample(3, random_state = 42)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
362514  19677589   21900299      32500        32500          32500.0   
288564  29755527   32278795      11000        11000          11000.0   
213591   1343334    1588314      30000        30000          30000.0   

              term  int_rate  installment grade sub_grade  \
362514   60 months     14.99       773.01     C        C5   
288564   60 months     20.99       297.53     E        E4   
213591   36 months     14.65      1034.83     C        C2   

                         emp_title emp_length home_ownership  annual_inc  \
362514  Licensed vocational nurse    < 1 year       MORTGAGE     65000.0   
288564               shift manager  10+ years       MORTGAGE     60000.0   
213591         FastMed urgent care     1 year       MORTGAGE    175000.0   

       verification_status issue_d loan_status pymnt_plan  \
362514            Verified  Jul-14  Fully Paid          n   
288564            Verified  Oct-14     Current          n   
213591            Verified  Jun-12  Fully Paid          n   

                                                      url  \
362514  https://www.lendingclub.com/browse/loanDetail....   
288564  https://www.lendingclub.com/browse/loanDetail....   
213591  https://www.lendingclub.com/browse/loanDetail....   

                                                     desc             purpose  \
362514                                                NaN    home_improvement   
288564                                                NaN         credit_card   
213591    Borrower added on 06/06/12 > I am seeking a ...  debt_consolidation   

                          title zip_code addr_state    dti  delinq_2yrs  \
362514         Home improvement    950xx         CA   9.32          0.0   
288564  Credit card refinancing    472xx         IN  24.19          1.0   
213591       Debt Consolidation    852xx         AZ  13.34          0.0   

       earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
362514           Apr-02             0.0                     NaN   
288564           Aug-03             0.0                     3.0   
213591           Mar-02             0.0                    41.0   

        mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  \
362514                     NaN       7.0      0.0       1208         4.3   
288564                     NaN      18.0      0.0      13129        54.7   
213591                     NaN       9.0      0.0       1906        22.4   

        total_acc initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
362514       20.0                   f       0.00           0.00     38213.56   
288564       43.0                   w    9214.67        9214.67      4450.12   
213591       38.0                   f       0.00           0.00     32388.64   

        total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
362514         38213.56         32500.00        5713.56                 0.0   
288564          4450.12          1785.33        2664.79                 0.0   
213591         32388.64         30000.00        2388.64                 0.0   

        recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
362514         0.0                      0.0       Oct-15         27391.42   
288564         0.0                      0.0       Jan-16           297.53   
213591         0.0                      0.0       Jan-13         26181.92   

       next_pymnt_d last_credit_pull_d  collections_12_mths_ex_med  \
362514          NaN             Oct-15                         0.0   
288564       Feb-16             Jan-16                         0.0   
213591          NaN             Mar-13                         0.0   

        mths_since_last_major_derog  policy_code application_type  \
362514                          NaN            1       INDIVIDUAL   
288564                          NaN            1       INDIVIDUAL   
213591                          NaN            1  

In [5]:
df.shape

(466285, 74)

In [6]:
list_item = []
for col in df.columns :
    list_item.append([col, df[col].dtype, df[col].count(), df[col].isna().sum(),
                      100 * df[col].isna().sum() / len(df[col]),
                      df[col].min() if df[col].dtype in ['int64', 'float64'] else '-',
                      df[col].max() if df[col].dtype in ['int64', 'float64'] else '-',
                      df[col].nunique(), df[col].unique()])
desc_df = pd.DataFrame(data = list_item, columns = ['feature', 'data_type', 'fill_num', 'null_num', 'null_percent', 'min_value', 'max_value', 'unique_num', 'unique_sample'])
desc_df

feature data_type  fill_num  null_num  null_percent  \
0                            id     int64    466285         0      0.000000   
1                     member_id     int64    466285         0      0.000000   
2                     loan_amnt     int64    466285         0      0.000000   
3                   funded_amnt     int64    466285         0      0.000000   
4               funded_amnt_inv   float64    466285         0      0.000000   
5                          term    object    466285         0      0.000000   
6                      int_rate   float64    466285         0      0.000000   
7                   installment   float64    466285         0      0.000000   
8                         grade    object    466285         0      0.000000   
9                     sub_grade    object    466285         0      0.000000   
10                    emp_title    object    438697     27588      5.916553   
11                   emp_length    object    445277     21008      4.505399   
12               home_ownership    object    466285         0      0.000000   
13                   annual_inc   float64    466281         4      0.000858   
14          verification_status    object    466285         0      0.000000   
15                      issue_d    object    466285         0      0.000000   
16                  loan_status    object    466285         0      0.000000   
17                   pymnt_plan    object    466285         0      0.000000   
18                          url    object    466285         0      0.000000   
19                         desc    object    125981    340304     72.981975   
20                      purpose    object    466285         0      0.000000   
21                        title    object    466264        21      0.004504   
22                     zip_code    object    466285         0      0.000000   
23                   addr_state    object    466285         0      0.000000   
24                          dti   float64    466285         0      0.000000   
25                  delinq_2yrs   float64    466256        29      0.006219   
26             earliest_cr_line    object    466256        29      0.006219   
27               inq_last_6mths   float64    466256        29      0.006219   
28       mths_since_last_delinq   float64    215934    250351     53.690554   
29       mths_since_last_record   float64     62638    403647     86.566585   
30                     open_acc   float64    466256        29      0.006219   
31                      pub_rec   float64    466256        29      0.006219   
32                    revol_bal     int64    466285         0      0.000000   
33                   revol_util   float64    465945       340      0.072917   
34                    total_acc   float64    466256        29      0.006219   
35          initial_list_status    object    466285         0      0.000000   
36                    out_prncp   float64    466285         0      0.000000   
37                out_prncp_inv   float64    466285         0      0.000000   
38                  total_pymnt   float64    466285         0      0.000000   
39              total_pymnt_inv   float64    466285         0      0.000000   
40              total_rec_prncp   float64    466285         0      0.000000   
41                total_rec_int   float64    466285         0      0.000000   
42           total_rec_late_fee   float64    466285         0      0.000000   
43                   recoveries   float64    466285         0      0.000000   
44      collection_recovery_fee   float64    466285         0      0.000000   
45                 last_pymnt_d    object    465909       376      0.080637   
46              last_pymnt_amnt   float64    466285         0      0.000000   
47                 next_pymnt_d    object    239071    227214     48.728567   
48           last_credit_pull_d    object    466243        42      0.009007   
49   collections_12_mths_ex_med   float64    466140       145      0.03109

In [7]:
desc_df.data_type.value_counts()

data_type
float64    46
object     22
int64       6
Name: count, dtype: int64

In [8]:
desc_df[desc_df.null_percent > 20].shape[0]

22

In [9]:
desc_df[(desc_df.null_percent <= 20) & (desc_df.null_percent > 0)].shape[0]

18

In [10]:
df_nums = df.select_dtypes(include = ['int64', 'float64'])
df_cats = df.select_dtypes(include = 'object')

In [11]:
from IPython.display import display, HTML

display(HTML('<center><b><h3>Statistics Description of Dataset</h3></b></center><center><b><h5>Numerical Features</h5></b></center>'))
display(df_nums.describe())

id     member_id      loan_amnt    funded_amnt  \
count  4.662850e+05  4.662850e+05  466285.000000  466285.000000   
mean   1.307973e+07  1.459766e+07   14317.277577   14291.801044   
std    1.089371e+07  1.168237e+07    8286.509164    8274.371300   
min    5.473400e+04  7.047300e+04     500.000000     500.000000   
25%    3.639987e+06  4.379705e+06    8000.000000    8000.000000   
50%    1.010790e+07  1.194108e+07   12000.000000   12000.000000   
75%    2.073121e+07  2.300154e+07   20000.000000   20000.000000   
max    3.809811e+07  4.086083e+07   35000.000000   35000.000000   

       funded_amnt_inv       int_rate    installment    annual_inc  \
count    466285.000000  466285.000000  466285.000000  4.662810e+05   
mean      14222.329888      13.829236     432.061201  7.327738e+04   
std        8297.637788       4.357587     243.485550  5.496357e+04   
min           0.000000       5.420000      15.670000  1.896000e+03   
25%        8000.000000      10.990000     256.690000  4.500000e+04   
50%       12000.000000      13.660000     379.890000  6.300000e+04   
75%       19950.000000      16.490000     566.580000  8.896000e+04   
max       35000.000000      26.060000    1409.990000  7.500000e+06   

                 dti    delinq_2yrs  inq_last_6mths  mths_since_last_delinq  \
count  466285.000000  466256.000000   466256.000000           215934.000000   
mean       17.218758       0.284678        0.804745               34.104430   
std         7.851121       0.797365        1.091598               21.778487   
min         0.000000       0.000000        0.000000                0.000000   
25%        11.360000       0.000000        0.000000               16.000000   
50%        16.870000       0.000000        0.000000               31.000000   
75%        22.780000       0.000000        1.000000               49.000000   
max        39.990000      29.000000       33.000000              188.000000   

       mths_since_last_record       open_acc        pub_rec     revol_bal  \
count            62638.000000  466256.000000  466256.000000  4.662850e+05   
mean                74.306012      11.187069       0.160564  1.623020e+04   
std                 30.357653       4.987526       0.510863  2.067625e+04   
min                  0.000000       0.000000       0.000000  0.000000e+00   
25%                 53.000000       8.000000       0.000000  6.413000e+03   
50%                 76.000000      10.000000       0.000000  1.176400e+04   
75%                102.000000      14.000000       0.000000  2.033300e+04   
max                129.000000      84.000000      63.000000  2.568995e+06   

          revol_util      total_acc      out_prncp  out_prncp_inv  \
count  465945.000000  466256.000000  466285.000000  466285.000000   
mean       56.176947      25.064430    4410.062342    4408.452258   
std        23.732628      11.600141    6355.078769    6353.198001   
min         0.000000       1.000000       0.000000       0.000000   
25%        39.200000      17.000000       0.000000       0.000000   
50%        57.600000      23.000000     441.470000     441.380000   
75%        74.700000      32.000000    7341.650000    7338.390000   
max       892.300000     156.000000   32160.380000   32160.380000   

         total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
count  466285.000000    466285.000000    466285.000000  466285.000000   
mean    11540.686220     11469.892747      8866.014657    2588.677225   
std      8265.627112      8254.157579      7031.687997    2483.809661   
min         0.000000         0.000000         0.000000       0.000000   
25%      5552.125349      5499.250000      3708.560000     957.280000   
50%      9419.250943      9355.430000      6817.760000    1818.880000   
75%     15308.158460     15231.310000     12000.000000    3304.530000   
max     57777.579870     57777.580000     35000.030000   24205.620000   

       total_rec_late_fee     recoveries  collection_recovery_fee  \
count       466285.000000  

In [12]:
display(HTML('<center><b><h3>Statistics Description of Dataset</h3></b></center><center><b><h5>Categorical Features</h5></b></center>'))
display(df_cats.describe().T)

count  unique  \
term                 466285       2   
grade                466285       7   
sub_grade            466285      35   
emp_title            438697  205475   
emp_length           445277      11   
home_ownership       466285       6   
verification_status  466285       3   
issue_d              466285      91   
loan_status          466285       9   
pymnt_plan           466285       2   
url                  466285  466285   
desc                 125981  124435   
purpose              466285      14   
title                466264   63098   
zip_code             466285     888   
addr_state           466285      50   
earliest_cr_line     466256     664   
initial_list_status  466285       2   
last_pymnt_d         465909      98   
next_pymnt_d         239071     100   
last_credit_pull_d   466243     103   
application_type     466285       1   

                                                                   top    freq  
term                                                         36 months  337953  
grade                                                                B  136929  
sub_grade                                                           B3   31686  
emp_title                                                      Teacher    5399  
emp_length                                                   10+ years  150049  
home_ownership                                                MORTGAGE  235875  
verification_status                                           Verified  168055  
issue_d                                                         Oct-14   38782  
loan_status                                                    Current  224226  
pymnt_plan                                                           n  466276  
url                  https://www.lendingclub.com/browse/loanDetail....       1  
desc                                                                       234  
purpose                                             debt_consolidation  274195  
title                                               Debt consolidation  164075  
zip_code                                                         945xx    5304  
addr_state                                                          CA   71450  
earliest_cr_line                                                Oct-00    3674  
initial_list_status                                                  f  303005  
last_pymnt_d                                                    Jan-16  179620  
next_pymnt_d                                                    Feb-16  208393  
last_credit_pull_d                                              Jan-16  327699  
application_type                                            INDIVIDUAL  466285

In [13]:
for column in df.columns :
    print(f"Data terbanyak pada kolom {column} :")
    print(df[column].value_counts().head(5))
    print("\n")

Data terbanyak pada kolom id :
id
1077501     1
26180404    1
26219322    1
26239577    1
26290537    1
Name: count, dtype: int64


Data terbanyak pada kolom member_id :
member_id
1296599     1
28653081    1
28692177    1
28702376    1
28763241    1
Name: count, dtype: int64


Data terbanyak pada kolom loan_amnt :
loan_amnt
10000    33023
12000    25519
15000    23486
20000    22759
35000    16596
Name: count, dtype: int64


Data terbanyak pada kolom funded_amnt :
funded_amnt
10000    32930
12000    25426
15000    23370
20000    22571
35000    16458
Name: count, dtype: int64


Data terbanyak pada kolom funded_amnt_inv :
funded_amnt_inv
10000.0    28683
12000.0    21732
15000.0    19473
20000.0    18070
8000.0     13232
Name: count, dtype: int64


Data terbanyak pada kolom term :
term
 36 months    337953
 60 months    128332
Name: count, dtype: int64


Data terbanyak pada kolom int_rate :
int_rate
12.99    15846
10.99    14837
15.61    13074
8.90     11424
16.29    10587
Name: count, d

### Beberapa analisa dari dataset yang ditemukan antara lain :
- Terdapat 466285 baris dan 74 kolom pada dataset ini.
- Ada 22 kolom kategorikal dan 52 kolom numerikal.
- Beberapa kolom memiliki nilai yang *unique*, antara lain `id`, `member_id` dan `url`. Untuk kolom-kolom tersebut perlu dilakukan *drop column*.
- Kolom `policy_code` dan `application_type` memiliki jumlah *unique* hanya satu, sehingga perlu dilakukan *drop column*.
- Ada 22 kolom yang memiliki data kosong lebih dari 20 persen dari keseluruhan data, lebih baik dilakukan *drop column* juga.
- Ada 18 kolom yang memiliki data kosong di bawah 20 persen dari keseluruhan data, untuk ini dilakukan *remove missing values* tersebut.